pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,346 @@
+}
+
+func (h *HistogramFunctionOverInstantVector) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+	if len(h.remainingGroups) == 0 {
+		// No more groups left.
+		return types.InstantVectorSeriesData{}, types.EOS
+	}
+
+	// Determine next group to return
+	thisGroup := h.remainingGroups[0]
+	h.remainingGroups = h.remainingGroups[1:]
+	defer func() {
+		// Reset the group before returning to the pool
+		thisGroup.groupedMetricName = ""
+		thisGroup.pointBuckets = nil
+		thisGroup.nativeHistograms = nil
+		thisGroup.remainingSeriesCount = 0
+		bucketGroupPool.Put(thisGroup)
+	}()
+
+	// Iterate through inner series until the desired group is complete
+	if err := h.accumulateUntilGroupComplete(ctx, thisGroup); err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	if len(h.remainingGroups) == 0 {
+		// We no longer need the bucketValues strings
+		types.StringSlicePool.Put(h.bucketValues, h.memoryConsumptionTracker)
+	}
+
+	return h.computeOutputSeriesForGroup(ctx, thisGroup)
+}
+
+// accumulateUntilGroupComplete gathers all the series associated with the given bucketGroup
+// As each inner series is selected, it is added into its respective groups.
+// This means a group other than the one we are focused on may get completed first, but we
+// continue until the desired group is ready.
+func (h *HistogramFunctionOverInstantVector) accumulateUntilGroupComplete(ctx context.Context, g *bucketGroup) error {
+	for g.remainingSeriesCount > 0 {
+		s, err := h.inner.NextSeries(ctx)
+		if err != nil {
+			if errors.Is(err, types.EOS) {
+				return fmt.Errorf("exhausted series before all groups were completed: %w", err)
+			}
+			return err
+		}
+		thisSeriesGroups := h.seriesGroups[h.currentInnerSeriesIndex]
+
+		// Native histograms only ever go to their original labelset.
+		// Floats only ever go to their group.
+		// It is possible that a series has both floats and histograms.
+		// It is also possible that both series groups are the same.
+		// The conflict in points is then detected in computeOutputSeriesForGroup.
+		h.saveNativeHistogramsToGroup(s.Histograms, thisSeriesGroups[0])
+		h.saveFloatsToGroup(s.Floats, thisSeriesGroups[1])
+		h.currentInnerSeriesIndex++
+	}
+	return nil
+}
+
+// saveFloatsToGroup places each fPoint into a bucket with the upperBound set by the input series.
+func (h *HistogramFunctionOverInstantVector) saveFloatsToGroup(fPoints []promql.FPoint, g *bucketGroup) {
+	if len(fPoints) > 0 {
+		upperBound, err := strconv.ParseFloat(h.bucketValues[h.currentInnerSeriesIndex], 64)
+		if err != nil {
+			// The le label was invalid. Record it:
+			h.Annotations.Add(annotations.NewBadBucketLabelWarning(
+				h.innerSeriesMetricNames.GetMetricNameForSeries(h.currentInnerSeriesIndex),
+				h.bucketValues[h.currentInnerSeriesIndex],
+				h.inner.ExpressionPosition(),
+			))
+		} else {
+			for _, f := range fPoints {
+				pointIdx := h.timeRange.PointIndex(f.T)
+				g.pointBuckets[pointIdx] = append(
```

Comment:
Something to consider, might be something for a follow-up PR: what if we allocated `g.pointBuckets[pointIdx]` once based on the expected number of buckets? As it stands, we'll keep appending to `g.pointBuckets[pointIdx]`, which may require many expansions of the slice, with all the allocations and copying that entails.

We could assume that if there are any floats present at a point, then all buckets will be present at that point (which should hold true unless the bucket layout changes).

We could also then pre-sort the list of buckets by `upperBound`, and then directly write to the correct bucket, reducing / eliminating any shuffling required when sorting in `bucketQuantile`.

The only thing I'm not sure about is how we'd handle the case where some buckets aren't present (eg. because the bucket layout changed mid-query) - we'd need to keep track of which buckets are present somehow.

Reply:
It's a good idea, but I think there are some complicated edge cases as you point out. So I agree better for a followup PR.

Reply:
> Allocating g.pointBuckets[pointIdx] based on the maximum expected number of buckets, and using a pool for this, rather than expanding it. The majority of the time, this will be the right thing to do, and in the worst case, we'll allocate some buckets that won't be used.

I think I would want to benchmark whether we gain much here, and so I'd rather do this as a followup at this point if that's okay. I don't want this PR to go on forever.

> Pre-sorting the list and taking advantage of that to avoid sorting in bucketQuantile - I think this can be left for a later PR

Prometheus are open to exporting the functions, so we need to decide if it's worth having these potential optimisations or if the maintenance burden is higher than we'd like and we'd rather rely on the identical (ie exported) functions.

Reply:
Fair enough, happy for that to be a follow-up PR.

Reply:
Okay, ended up doing this and benchmarks (see below). We can still make the optimisation of sorting, but that won't change the memory which is the current problem for this function. (And see comment about using exported option)